### DISCUSSION TITLE
Add the ability to reorder the sidebar tabs

### DISCUSSION THREAD

**Discussion by MagicfTail** [runelite/runelite#15020] [Ideas]
I feel like the logout tab should be at the bottom, other people might want another order. Being able to reorder it would be nice.

**Comment by ShaunDreclin** [comment]
~~duplicate of #582~~

**Comment by deathbeam** [comment]
As Shaun said

**Comment by MagicfTail** [comment]
From my understanding, that issue is requesting the reordering of the ingame tiles. I would like to be able to reorder the clients tabs, like the Xp Tracker, Info, Kurend Library plugin and so on.

**Comment by deathbeam** [comment]
Ah, yes, sorry, misunderstood the issue. Calling button tab is very confusing :( Anyway, for the login/logout problem, there is #646 PR open currently.

**Comment by ShaunDreclin** [comment]
Ohh I misunderstood too, yeah being able to re-order the client's tab strip would be nice.

Possibly related to that, #449 (Hotkeys for hiding the sidebar and for opening each tab)

**Comment by MagicfTail** [comment]
No worries, i could have been more precise when describing my request. Fel free to reword anythign to make it more precise :)

**Comment by yoghurtis** [comment]
Yes please!

**Comment by RickertdeH** [comment]
This would be useful, especially with all the new external plugins available.

**Comment by wigglepop** [comment]
![image](https://user-images.githubusercontent.com/77034846/103753112-994f2b00-5045-11eb-8147-643e4c3b5312.png)
if the ability to re-arrange sidebars is not going to be put, please add a scrolling ability for those who don't play full screen. it's hard to put the client to full screen again and again just so we could open up a plugin that's at the bottom of the list

**Comment by jevonsp** [comment]
Is this ever going to be revisited? with the plugin hub it is a larger annoyance

**Comment by 420Lunacy** [comment]
I agree with everyone here, the ability to re-order the side bar is something that needs to happen. Especially seeing as how people are asking for it constantly because they are unable to see all their plugins.

**Comment by ghost** [comment]
Agreed, this needs to be revisited. I play full screen and I just hit the max icons.

**Comment by javenjefven** [comment]
scrolling or the ability to rearrange the icons would be excellent!

**Comment by Ch00chinat0r** [comment]
Still hoping this gets added to RuneLite

**Comment by Jbleezy** [comment]
How has this not been added? It seems simple but maybe I'm wrong.

**Comment by psbj** [comment]
I've been wondering why RL never had this, really should be added!

**Comment by Troy-MC** [comment]
To be honest I'm surprised I can't find an existing plugin that does this.

I noticed that some existing plugins offer the option to customise their icon's relative position on the Runelite sidebar, such as [this one](https://github.com/Mrnice98/BossingInfo).

This tells me there should be a degree of control over the sidebar offered by the plugin API.
I'll have a look at making a plugin that allows users to customise the relative position of all the icons, since it's something I'd like to use too. 

If I don't write back here within a month of this post you can safely assume I forgot about this.

**Comment by Osamu01** [comment]
I came here to support this. Pretty avid runelite user and would benefit off being able to find my most used plugins easier.

**Comment by jevonsp** [comment]
> If I don't write back here within a month of this post you can safely assume I forgot about this.

pinging you because I'd really like to get this feature implemented 😊

**Comment by K3irara** [comment]
will this be added? also a way to show text labels?

**Comment by swirle13** [comment]
@Troy-MC pinging you again as you said if you didn't come back in a month (back in March, 9 months ago), we can assume you forgot about it.

I'm a dev and I have experience with different languages and platforms but I haven't worked on anything with runelite or openGL. I'd be willing to take a crack at this too, but I'd be starting from scratch: how to contribute, what do to, how to work on runelite, etc. So it'd take me a while, amongst all the other projects I'm working on lol.

Tagging someone within runelite to get more visibility on this, @raiyni since you were the one that converted this issue into a discussion, which to me, just makes it less noticeable and takes the power away from users who want this to be implemented. Can you see all the comments and upvotes for everyone who wants this? Is there any way to put this on the backlog?

**Reply by raiyni** [reply]
Discussions are a better format for feature requests and ideas than creating issues. 

This is a feature that has been heavily talked about in relation to the flatlaf uplift. Anyone can work on most issues, they don't need permission or assignment.

**Reply by Nealon01** [reply]
This continues to be a pretty significant issue once you install enough plugins... Only getting worse.

Would be lovely to have the option to group plugins into folders and show/hide individual ones.

**Reply by swirle13** [reply]
Alright y'all, I'm back and I have delivered! I finally learned enough by contributing to plugins to take my own shot at this. Spent a few days refining the behavior, including:

- allowing toggling between default and custom plugin order
- ensuring new plugins are added at a reasonable position if custom order is enabled
- persisting between sessions
- dragging without switching plugin panels
- maintaining plugin order if a plugin is turned off and back on later (unless you reset your custom order)
- toggle to require a hotkey to drag icons (prevents accidental drags at the worst times)
- hot key selector (your choice of hotkey)
- plus some fancy UI tricks to make it feel polished (not hover highlighting tabs when dragging)

I made a couple different iterations:

1. Slide and Stop
    - mimics the other drag and drop behavior in the client (these aren't the same component types so the behavior had to be replicated as close as able) which looked nice, but has some performance issues when dragging active plugins with large amounts of elements, e.g. Configuration and World Switcher plugins, as you can see at the end of the video.
2. Drag and Drop
	- A more performant option with a dragged icon, some transparency to see what you're dragging over, and a nice insertion indicator. This implementation works better with the UI element these are (JTabbedPane)

Slide and Stop | Drag and Drop
--- | ---
![Slide_and_Stop](https://github.com/user-attachments/assets/6a0526c6-79a1-4afc-93ef-2d6ed8103df6) | ![Drag_and_Drop](https://github.com/user-attachments/assets/61693b3f-df83-4c46-8630-a4545f7dec5d)

**Reply by Infinitay** [reply]
Drag and drop looks perfect and seems more performative for sure. If you created this as a plugin, it'd be wonderful if you'd publish it. Also, any chance you could add support for (un)hiding certain panels? Maybe you can work with @Wynadorn and adopt his panel visibility feature (https://github.com/runelite/runelite/discussions/15020#discussioncomment-15233894)

**Reply by ShaunDreclin** [reply]
The chosen one has blessed us

Please god let this be merged

**Reply by swirle13** [reply]
@Infinitay After working on this write-up, I reviewed Wynadorn's changes and I liked the hidden panel feature. I also had seen @roskabytes comment about accessibility. I'm not quite sure how to allow this, but I don't think either are impossible.

As it is, the PR will be about 10 files changed and 2200 lines added 😬
<img width="594" height="44" alt="image" src="https://github.com/user-attachments/assets/ec38d32d-939a-4688-95bb-0912b2bc4f5f" />

500 lines are a new UI component to mirror the existing dragging behavior but for tabs (which are the plugin icon type), 1200 lines for unit tests, the plugin itself is about 400 lines, and changes to support the plugin are about 100 lines.

Praying the maintainers don't see the size of the PR and immediately ignore it Prayge

So extra features would be good but probably have to come afterwards if I want the maintainers to actually review it

**Comment by MuchoBobito** [comment]
I take it there is no new news on this as of yet?

**Reply by Nealon01** [reply]
Of course not.  Far too easy/impactful.

EDIT: marked as spam 😂

**Comment by Blorog1** [comment]
It's been six years and this hasn't been implemented. Amazing.

**Comment by 420Lunacy** [comment]
At this point I feel like they are deliberately not even replying to this
email link saying yay/nay. I think at this point since it's been out for so
long and is such a sick quality of life upgrade, it's likely they either
felt the task was beneath their level and wanted to work on "harder things"
or they are just busy and unable to find the importance of stopping their
current project to try and trouble shoot (any) fix for it all.  since the
shit they are busy takes up most of them.

**Comment by tylerwgrass** [comment]
This hasn't been closed so I'd assume that means they are open to the feature being added. There are limited maintainers and a lot of issues. Just because something seems like an easy change doesn't mean it is in implementation and there are other considerations beyond just making it work.

As mentioned already, anyone is welcome to try and contribute this if they so choose.

**Reply by Varietyz** [reply]
Challenge accepted!

**Comment by Osamu01** [comment]
Would love to  see someone tackle this given the recent discussions - hats off to you fancy pants developers!

**Comment by Varietyz** [comment]
![first_tab_reorder_test](https://github.com/user-attachments/assets/7a7ef4f0-8ca5-463d-8170-4f19f61ce722)
Working on Saving and loading now!

**Reply by Nealon01** [reply]
Give me your paypal you fucking legend, I'll buy you some coffee.

6 years of "but it's too hard", and this legend does it in 45 minutes.

**Reply by ShaunDreclin** [reply]
Wow this looks awesome! If you're able to could you also implement @wigglepop 's suggestion of scrolling the tabs?

**Reply by Varietyz** [reply]
Fella's, ive done it.

**Reply by Varietyz** [reply]
doing final testing before commiting :)

**Reply by Varietyz** [reply]
https://github.com/runelite/runelite/pull/18406

**Reply by Saelem** [reply]
Is this still happening, we really need this, thanks Varietyz for making it happen! 👯

**Reply by Varietyz** [reply]
> Is this still happening, we really need this, thanks Varietyz for making it happen! 👯

Its been tougher then i assumed at first with utilizing runelite's current config system.

Also been having some irl issues which have cut my time shorter than I would personally like.
I havent given up on it and hope to have a decent enough state to submit a pr before the end of 2024.

**Reply by Vequis-uk** [reply]
Hope you are ok man, really appreciate any work on Runelite it really is amazing <3

**Reply by idmoore** [reply]
@Varietyz Keep moving forward!! How's it going? :)

**Comment by ShaunDreclin** [comment]
This feature is still really needed, my sidebar is a mess!

**Reply by Sharpienero** [reply]
> This feature is still really needed, my sidebar is a mess!

Uninstall some plugins

**Reply by notheowner** [reply]
![image](https://github.com/user-attachments/assets/62023d0a-ff9c-468c-b09c-c8074acc737e)

**Comment by Eiziv** [comment]
Is this happening? I need this in my life

**Comment by Wynadorn** [comment]
Oh boy, I didn't even realize this discussions thread existed.
I do know that in the past there've been quite a couple of drafts implementations to add this feature to runelite.

So to keep that tradition going here's mine 😶‍🌫️:

https://github.com/user-attachments/assets/d82bc10d-5fad-43d7-a3e0-56c31f9d8921

Jokes aside though, there's been quite some discussion about the way in which this feature should be implemented.
Instead of allowing the buttons to be re-ordered I've opted to split the buttons into 3 sets, pinned, sorted (default) and hidden.

The pinned buttons can be reordered, the sorted buttons follow the default ordering as it is implemented right now, and the hidden buttons appear in a pop-up menu.

It's a pretty rough implementation right now,
but if the core functionality is alright then I could look into cleaning it up and then submitting a PR.

**Reply by Wynadorn** [reply]
As for the implementation side, I've replaced the `JTabbedPane` in `ClientUI` with a `SortableJTabbedPane`, which is basically the UI element you see in the video. To the outside it's meant to behave the same as a `JTabbedPane`.

**Reply by Infinitay** [reply]
I like the feature you have that lets you still interact with panels within your hidden visibility _panel_

**Reply by roskabytes** [reply]
Hey @Wynadorn , this would be awesome. I have one suggestion though: relying on dragging is very inaccessible for anyone dealing with any temporary or chronic motor difficulties - and just not practical if users have many plugins. Could there be an alternative way to implement this, even if it's via json settings tweaks? Thanks

**Reply by Wynadorn** [reply]
An import export feature wouldn't be difficult to add.

However It's been a while, I don't think it's currently a priority for the maintainers to review this. Would take a small miracle for this to be merged soon.

**Reply by Oridjinn1980** [reply]
Really, really excited for this! Hope you don't forget about it like the last few people who gave this a try 😭 As my plugins list grows, so too do my sidebar icons, and it's getting to be a bit ridiculous. Having the ability to reorder them, and hide the ones I rarely use, would be a gamechanger.

**Comment by roskabytes** [comment]
Although Wynadorn is working on a plugin to address this, I'd still like to add a vote for reordering to be added as a core Runelite functionality.

**Reply by Wynadorn** [reply]
Nah its not a plugin, this has to be a core feature, which is also why it's a difficult topic. Any changes would apply to all users.

It looks like a plugin, but that's just an enable / disable switch, similar to other default plugins.

**Comment by hauber794** [comment]
Would love to see this get implemented!

**Comment by Eiziv** [comment]
Id also love to have a sidebar on each side of my client

**Comment by gillavarmkorv-netizen** [comment]
Sorry to crash your idea but it's been requested multiple times and has to my knowledge been considered a non-fix issue.
Unless the core team has changed their views this functionality is unlikely to be implemented.
#3277
#1641
#13970

**Reply by Oridjinn1980** [reply]
None of the issues you linked are about reordering the sidebar, they're about moving the sidebar location altogether. Additionally they state that the "core team" won't work on it because it is not an issue, but if a PR is opened with the functionality added, it could be merged after review- although my faith in that is somewhat tempered by the fact that a PR #18262 adding the feature has been open for almost 2 years with seemingly no comment or even a glance from maintainers.

**Comment by Eiziv** [comment]
More hyped for this than raids 4

**Comment by 420Lunacy** [comment]
Maybe if everyone comments again about wanting it to get approved, it will push them to actually do it.

Sent via the Samsung Galaxy S24, an AT&T 5G smartphone
Get Outlook for Android<https://aka.ms/AAb9ysg>

 Stephanie Howell
________________________________

**Comment by swirle13** [comment]
Wanted to leave an updated comment of mine (mine is in a top level comment that is now hidden here).

I've been working with the RL discord and getting feedback from Adam (creator of RL and head dude) and other RL contributors. Sounds like they're all on board and through refinement, here's where we're at now:

https://github.com/user-attachments/assets/dcfd04b5-9ecc-40e2-9e30-793e39ed9d1d

The feature has been moved into the core runelite code, no longer a 1st party plugin. Custom ordering is always enabled. Drag uses the Runelite-wide UI drag key (Alt/Option by default). Right click "Reset" popup menu to allow resetting back to the default order, while still respecting any plugins that may have set their plugin icon to have a right click popup menu.

Future possible additions:

- Confirmation screen for resetting order
- Hiding icons (visually turned off or hidden in a drawer) + "reset visibility" popup menu entry
- Loading plugin order via import popup menu entry (helps accessibility). Would necessitate creation of "export" popup menu entry too.

Edit: added confirmation screen but can't add video, too large

Open PR can be found here: https://github.com/runelite/runelite/pull/19910

**Reply by Nealon01** [reply]
I demand you let me buy you a coffee once this is live.  You're actually pulling it off.  8 years later and you're succeeding where countless others have given up (myself included).  Thank you sir.  Godspeed.

**Reply by Nealon01** [reply]
Oh and I think the option for folders for groups of plugins (infrequently used, or related) would be a great addition too.  I have 2 columns of icons and the hidden would be nice, but better organization/icons for those folders would be huge.  No rush or expectations obviously, and I'd be happy to take a stab at that myself once the groundwork is there to build from.

**Reply by ShaunDreclin** [reply]
Not sure if it'd be out of scope, but the ability to hide unwanted sidebar tabs (either completely or stuffing them into a subfolder of some sort) would still be hugely beneficial as well.

**Reply by swirle13** [reply]
Latest update (I believe this is very nearly feature complete, for the time being):

Couldn't help myself, added proper RuneLite profile loading, allowing loading of different custom orders for different profiles. Also removed unit tests for smaller PR. Recording of behavior attached:

https://github.com/user-attachments/assets/9af7983b-b722-4f8e-9e28-cd813dbee0fe

Actions performed:
- reset order for profile Desktop
- dragged Gauntlet Map plugin to index 1
- switched to profile Macbook (has default order except Bot Detector at index 1)
- Compared against screenshot of state before switchign to Desktop profile
- validated Macbook profile custom order successfully loaded (plagued by bugs of off-by-one plugin order or ghost tabs of plugins not installed still staying)
- Switch back to Desktop profile
- Validate against screenshot of Desktop plugin order

**Comment by swirle13** [comment]
I wrote up a guide on how to use my sidebar changes for those who aren't familiar enough with runelite development to build it themselves. Looking for some volunteers to test the guide out as well as the sidebar changes to gather feedback.

https://gist.github.com/swirle13/454add1938876fb0bb971d16c9b69cae

CC'ing anyone that showed interest: @420Lunacy @Oridjinn1980 @Nealon01 @Tom-Cox-1 @Eiziv @hauber794 @ShaunDreclin @Infinitay @Osamu01 @jevonsp @Troy-MC @Ch00chinat0r @Jbleezy @psbj @javenjefven @wigglepop @RickertdeH @yoghurtis @MagicfTail 

Also, creator/author of Runelite said this today 👀 
<img width="624" height="101" alt="image" src="https://github.com/user-attachments/assets/4265fe4a-bf3e-44ec-8a5e-1238456523ae" />

**Reply by Nealon01** [reply]
Happy to help, would have to be Monday though.

**Reply by swirle13** [reply]
> Happy to help, would have to be Monday though.

Awesome. No rush. The PR has been done and ready for a few weeks. I was trying to find a way to be able to use the changes myself more easily and figured I'd write a guide for everyone else interested in trying it.

**Reply by Oridjinn1980** [reply]
Been loving the reordered sidebar! I've been using your fork since you posted about it here- so I guess about 2 weeks now? 

It's worked well in my testing, but unfortunately between commits the order reset to default and I lost my presets. It was right when you swapped to supporting profiles though, so that could be it too. Still need to do some more testing regarding plugins that support setting their position via gui and actually using profiles, but other than that it has been great! 

If possible, I would love to be able to create something like folders or categories so I could group the sidebar buttons together and hide / show them as necessary to help clean everything up and make it neater. Not sure how possible this would be though.

Really appreciate your hard work on this, and the commitment to get it merged! Needed something like this for so long

**Reply by swirle13** [reply]
Good news everyone! professor.gif

Adam has started work on implementing the sidebar reorder logic himself. He started with [tylerwgrass's PR](https://github.com/runelite/runelite/pull/18720/changes#diff-f190fc8fab40fd4e63e1744bda989d8ebff025ca58e3a6053a6a3c73ae69d5b5) for it and refactored a lot of the logic to be much nicer/quicker. he's working on the fundamentals and isn't at the UX part, which is where I'm hoping I can show him how mine behaves and what issues I encountered, e.g. tabs swapping when dragging the icons, drop tab indicator, dragged icon, right click reset behavior, etc. So my PR probably won't be merged, but I'm hoping some of my code can live on in Adam's changes.

But this is a massively strong sign that this is probably going to finally be added to runelite. I'll bring him the concerns/suggestions I've seen in here for UX, e.g. accessibility for users who can't drag icons, ability to hide/unhide unused icons, group/categorize icons, etc.

**Reply by hauber794** [reply]
@swirle13 Have you heard any update on progress? What you did is amazing work.

**Reply by swirle13** [reply]
> @swirle13 Have you heard any update on progress? What you did is amazing work.

Basically, my PR won't be merged because Adam said I did a bunch of things weird, so he took the code in and did it his way, then abex told him he did it poorly too, so abex started it and took it over and now it's just stalled out while abex's attention is elsewhere 🙃
